### PR TITLE
[JENKINS-30895] - Adding TestNG plugin support

### PR DESCRIPTION
**Pull request by daveLC**
Adding support for the TestNG plugin to allow publishing of TestNG reports.
This work was originally done by noroutine together with work to add support for the jenkins ci game:
https://github.com/jenkinsci/job-dsl-plugin/pull/501

This new pull request contains just the TestNG plugin changes.

### CODE REVIEW

**Comment by daveLC** [comment]
Not sure why this is failing on jenkins, it passes locally after removing the whitespace

**Comment by daspilker** [comment]
There was a whitespace problem on the master branch. If you rebase to master, it should be OK. See https://github.com/jenkinsci/job-dsl-plugin/commit/3a5e920b777a6c819a3367a22c855a5282432bec.

**Comment by daveLC** [comment]
Hmm, still failing on line 245 of the PublisherContext
https://jenkins.ci.cloudbees.com/job/plugins/job/job-dsl-plugin/ws/job-dsl-core/build/reports/codenarc/main.html#javaposse/jobdsl/dsl/helpers/publisher
But this line does not have trailing whitespace and it passes locally

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/TestNGDataPublishersContext.groovy [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package javaposse.jobdsl.dsl.helpers.publisher
+
+import javaposse.jobdsl.dsl.AbstractContext
+import javaposse.jobdsl.dsl.JobManagement
+
+class TestNGDataPublishersContext extends AbstractContext {
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/TestNGDataPublishersContext.groovy]: do not add empty context classes

File: job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/ArchiveTestNGContext.groovy [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    boolean escapeExceptionMessages = true
+    boolean showFailedBuildsInTrendGraph = false
+    boolean markBuildAsUnstableOnSkippedTests = false
+    boolean markBuildAsFailureOnFailedConfiguration = false
+
+    ArchiveTestNGContext(JobManagement jobManagement) {
+        testDataPublishersContext = new TestNGDataPublishersContext(jobManagement)
+    }
+
+    void escapeTestDescription(boolean escape = true) {
```
Comment (daspilker) [inline job-dsl-core/src/main/groovy/javaposse/jobdsl/dsl/helpers/publisher/ArchiveTestNGContext.groovy]: add GroovyDoc for each method

File: job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy [hunk 3]
```diff
@@ -538,6 +538,42 @@ class PublisherContextSpec extends Specification {
         1 * jobManagement.requirePlugin('xunit')
     }
 
+	def 'call testng archive with all args'() {
```
Comment (daspilker) [inline job-dsl-core/src/test/groovy/javaposse/jobdsl/dsl/helpers/publisher/PublisherContextSpec.groovy]: fix the indentation

**Comment by daspilker** [comment]
CodeNarc passes for me too. I will take care of that when merging the branch.

**Comment by daveLC** [comment]
All seem to pass now, and have made the changes as requested

**Comment by daspilker** [comment]
I did not manage to configure any test data publishers for the TestNG plugin. And I could not find any implementation for them in the TestNG plugin sources, see https://github.com/jenkinsci/testng-plugin-plugin/blob/master/src/main/java/hudson/plugins/testng/Publisher.java. Are you sure that the TestNG plugin supports them?

**Comment by daveLC** [comment]
Oh ok, I'll remove that.

**Comment by daveLC** [comment]
Sorted now, taking it on from someone else isn't quite as straightforward as first thought!

**Comment by daspilker** [comment]
Thanks. I rebased your changes and merged them manually.

**Pull request closed without merging**